bots/github-push

```diff
@@ -0,0 +1,125 @@
+#! /usr/bin/python3
+
+# github-push -- Push the local branch to a GitHub PR, with benefits
+#
+# This tool pushes your local changes to a PR and also adds a comment
+# that summarizes what has changed.  The idea is to cement the actual
+# history of the PR in its comments, regardless of rebases, rewrites,
+# squashed fixups, etc.
+#
+# The tool finds the pull request number by looking through all open
+# pull requests on GitHub.  Once it has been found, the number is
+# cached in the Git config file to speed up subsequent invocations.
+
+import subprocess
+import argparse
+import sys
+import time
+
+from task import github
+
+def execute(*args):
+    try:
+        output = subprocess.check_output(args, universal_newlines=True)
+    except subprocess.CalledProcessError as ex:
+        sys.exit(ex.returncode)
+    return output
+
+def git(*args):
+    return execute("git", *args)
+
+def git_config(name):
+    try:
+        output = subprocess.check_output([ "git", "config", name ], universal_newlines=True)
+    except subprocess.CalledProcessError as ex:
+        return None
+    return output.strip()
+
+def main():
+    parser = argparse.ArgumentParser(description='Force push after a rewrite')
+    parser.add_argument('--repo', help="The GitHub repository to work with", default="cockpit-project/cockpit")
+    opts = parser.parse_args()
+
+    api = github.GitHub(repo=opts.repo)
+
+    local = git("rev-parse", "HEAD").strip()
+    remote = git("rev-parse", "@{u}").strip()
+
+    if local == remote:
+        sys.stderr.write("Nothing to push.\n")
+        return 0
+
+    ( tracking_remote, tracking_ref ) = git("rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{u}").strip().split("/")
+    tracking_url = git("remote", "get-url", tracking_remote).strip()
+
+    push_dest = tracking_remote
+    push_refspecs = local + ":" + tracking_ref
+    push_lease = tracking_ref + ":" + remote
+
+    pull_number = None
+    pull_config = git_config("pull.%s.%s" % (tracking_remote, tracking_ref))
+    if not pull_config:
+        # Find the pull request that matches the tracking branch
+        pulls = api.pulls()
+        for p in pulls:
+            ref = p["head"]["ref"]
+            url = "git@github.com:%s.git" % p["head"]["repo"]["full_name"]
+            if ref == tracking_ref and url == tracking_url:
+                pull_number = p["number"]
+                break
+        if not pull_number:
+            sys.stderr.write("No pull request found for %s/%s\n" % (tracking_remote, tracking_ref))
+            return 1
+        git("config", "pull.%s.%s" % (tracking_remote, tracking_ref), str(pull_number))
+    else:
+        pull_number=int(pull_config)
```

Comment:
It seems worth moving the determination of the pull number into a function, with some comments.

Reply:
The code has been refactored quite a lot now. How does it look now?